More tests and suggestions on default literal

jcouv opened the pull request:
@cston I'll stop by to discuss

cston commented on src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedDefaultTests.cs:
```diff
@@ -61,6 +61,155 @@ static void Main()
+";
+
+            var comp = CreateCompilationWithMscorlib(source, parseOptions: TestOptions.Regular7_1, options: TestOptions.DebugExe);
+            comp.VerifyDiagnostics(); // Should give an error. Follow-up issue: https://github.com/dotnet/roslyn/issues/18609
```
`using (null) { }` and `using (default(System.IDisposable)) { }` do not generate errors.

cston commented on src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedDefaultTests.cs:
```diff
@@ -61,6 +61,155 @@ static void Main()
+                //         F(async () => await default);
+                Diagnostic(ErrorCode.ERR_CantInferMethTypeArgs, "F").WithArguments("C.F<T>(System.Threading.Tasks.Task<T>)").WithLocation(8, 9)
+                );
+        }
```
Consider testing:
```
using System.Threading.Tasks;
class C
{
    static async Task<T> F<T>()
    {
        await Task.Delay(0);
        return default;
    }
}
```

cston commented:
LGTM

jcouv commented:
@dotnet/roslyn-compiler for review. This is mostly adding tests from discussion with Chuck so that I can merge "default" back into master.

cston commented on src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedDefaultTests.cs:
```diff
@@ -642,6 +691,93 @@ static void Main()
+";
+            var comp = CreateCompilationWithMscorlib(source, parseOptions: TestOptions.Regular7_1, options: TestOptions.DebugExe);
+            comp.VerifyDiagnostics();
+            CompileAndVerify(comp, expectedOutput: "reached");
```
Perhaps add `Console.WriteLine` to `operator==`.

gafter commented on src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedDefaultTests.cs:
```diff
@@ -193,310 +422,828 @@ static void Main()
+                // (14,17): error CS1031: Type expected
+                //         default();
+                Diagnostic(ErrorCode.ERR_TypeExpected, ")").WithLocation(14, 17),
+                // (6,17): error CS8119: The switch expression must be a value; found default.
```
I think the word default in this diagnostic should have some kind of quotes around it. It isn't a word of the sentence, it is referring to syntax in the source.

gafter commented on src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedDefaultTests.cs:
```diff
@@ -193,310 +422,828 @@ static void Main()
+                // (16,19): error CS1059: The operand of an increment or decrement operator must be a variable, property or indexer
+                //         int i = ++default;
+                Diagnostic(ErrorCode.ERR_IncrementLvalueExpected, "default"),
+                // (17,26): error CS0828: Cannot assign default to anonymous type property
```
Here too.

gafter commented on src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedDefaultTests.cs:
```diff
@@ -193,310 +422,828 @@ static void Main()
+";
+            var comp = CreateCompilationWithMscorlib(source, parseOptions: TestOptions.Regular7_1, options: TestOptions.DebugExe);
+            comp.VerifyDiagnostics(
+                // (11,13): error CS0023: Operator '!' cannot be applied to operand of type 'default'
```
Looking at the specification, this is incorrect:

> For an operation of the form `!x`, unary operator overload resolution (§13.4.4) is applied to select a specific operator implementation. The operand is converted to the parameter type of the selected operator, and the type of the result is the return type of the operator. Only one predefined logical negation operator exists:
> ``` c#
> bool operator !(bool x);
> ```
> This operator computes the logical negation of the operand: If the operand is true, the result is false. If the operand is false, the result is true.
> 
> Lifted (§13.4.8) forms of the unlifted predefined logical negation operator defined above are also predefined.

gafter commented on src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedDefaultTests.cs:
```diff
@@ -1168,12 +1984,34 @@ static void M(int x = default)
+    {
+        M();
+    }
+    static void M(System.Threading.CancellationToken x = default)
```
YES!

jcouv commented:
@cston This is ready for review.
Fixed EE scenario, added some tests, fixed unary operator case. It turns out the problem with `default` in interpolated strings and in `using` also repro'es with `null` (tracked by follow-up issue).

cston commented on src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs:
```diff
@@ -2180,7 +2180,7 @@ private BoundExpression BindUnaryOperatorCore(CSharpSyntaxNode node, string oper
 
-            bool isOperandTypeNull = (object)operand.Type == null;
+            bool isOperandTypeNull = operand.IsLiteralNull();
             if (isOperandTypeNull)
```
Are we reporting an error for an operand with no type, for instance `BoundMethodGroup`?

jcouv replied:
Yes, added test for method group and `null`.

cston commented on src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/LocalsTests.cs:
```diff
@@ -3766,6 +3766,65 @@ static void Main()
+                Assert.Equal(default(DkmEvaluationResultCategory), resultProperties.Category); // Not Data
+                Assert.Equal(default(DkmEvaluationResultAccessType), resultProperties.AccessType); // Not Public
+                Assert.Equal(default(DkmEvaluationResultStorageType), resultProperties.StorageType);
+                Assert.Equal(default(DkmEvaluationResultTypeModifierFlags), resultProperties.ModifierFlags); // Not Virtual
```
Please call `VerifyIL()`.

cston commented on src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/LocalsTests.cs:
```diff
@@ -3766,6 +3766,65 @@ static void Main()
+
+                var testData3 = new CompilationTestData();
+                context.CompileExpression("default", DkmEvaluationFlags.None, ImmutableArray<Alias>.Empty, out error, testData3);
+                Assert.Null(error);
```
What is the target type?

jcouv replied:
It's `object`.
The reason is that the EE adds a conversion:

```C#
        private static BoundStatement BindExpression(Binder binder, ExpressionSyntax syntax, DiagnosticBag diagnostics, out ResultProperties resultProperties)
        {
           ...
            var expressionType = expression.Type;
            if ((object)expressionType == null)
            {
                expression = binder.CreateReturnConversion(
                    syntax,
                    diagnostics,
                    expression,
                    RefKind.None,
                    binder.Compilation.GetSpecialType(SpecialType.System_Object));
                if (diagnostics.HasAnyErrors())
                {
                    resultProperties = default(ResultProperties);
                    return null;
                }
            }
```

cston replied:
Thanks. Consider adding `Assert.Equal(SpecialType.Object, method.ReturnType.SpecialType);` here in that case.

cston commented on src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/LocalsTests.cs:
```diff
@@ -3766,6 +3766,65 @@ static void Main()
         }
 
+        [Fact]
+        public void AssignDefaultToLocal()
```
Consider moving to `ExpressionCompilerTests`. `LocalsTests` was, at least initially, intended for testing `CompileGetLocals`.

cston commented on src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/SyntaxHelpers.cs:
```diff
@@ -210,7 +210,7 @@ private static InternalSyntax.ExpressionSyntax ParseDebuggerExpressionInternal(S
         {
             var source = SourceText.From(text);
-            using (var lexer = new InternalSyntax.Lexer(source, CSharpParseOptions.Default))
+            using (var lexer = new InternalSyntax.Lexer(source, CSharpParseOptions.Default.WithLanguageVersion(LanguageVersion.Latest)))
```
Please create a helper method for `ParseOptions`.

cston commented on src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/ExpressionCompilerTests.cs:
```diff
@@ -6205,5 +6205,86 @@ .maxstack  3
+}");
+
+                testData = new CompilationTestData();
+                context.CompileExpression("a = default;", DkmEvaluationFlags.None, ImmutableArray<Alias>.Empty, out error, testData);
```
Consider testing a non-expression statement (perhaps "int b = default;") to avoid having these succeed by falling back to expression parsing.

cston commented on src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/SyntaxHelpers.cs:
```diff
@@ -194,9 +194,11 @@ private static ExpressionSyntax ParseDebuggerExpression(string text, bool consum
             return expression.MakeDebuggerExpression(source);
         }
 
+        static CSharpParseOptions GetCSharpParseOptions() => CSharpParseOptions.Default.WithLanguageVersion(LanguageVersion.Latest);
```
Consider making this a `static readonly` field.

cston commented:
LGTM
